addons/mass_mailing/tests/test_mailing_retry.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.mass_mailing.tests.common import MassMailCommon
+from odoo.addons.base.tests.test_ir_cron import CronMixinCase
+from odoo.tests.common import users
+
+class TestMailingRetry(MassMailCommon, CronMixinCase):
+
+    @users('user_marketing')
+    def test_mailing_retry_immediate_trigger(self):
+        default_list_ids = self.env['mailing.list'].with_context(self._test_context).create({
+            'name': 'ListC',
+            'contact_ids': [
+                (0, 0, {'name': 'test', 'email': 'test@odoo.com'}),
+            ]
+        }).ids
+        mailing = self.env['mailing.mailing'].create({
+            'name': 'TestMailing',
+            'subject': 'Test',
+            'mailing_type': 'mail',
+            'body_html': '<div>Hello</div>',
+            'mailing_model_id': self.env['ir.model']._get('res.partner').id,
+            'contact_list_ids': default_list_ids,
+        })
+        mailing.action_launch()
+
+        # Send the mailing, which leads to RETRY because outgoing mail server is not set
+        self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo().method_direct_trigger()
```

Comment:
I don't think I like the fact that we are "expecting the outgoing mail server to fail".
The common test classes could be changed later to mock successful mail sending and then this test would not work anymore.

Instead we could force it to fail by maybe mocking the "send" method on "mail.mail".

Would be something like:

```py
# force email sending to fail to test our retry mechanism
def patched_mail_mail_send(mail_records, auto_commit=False, raise_exception=False):
    mail_records.write({'state': 'exception', 'failure_reason': 'forced_failure'})
    
with patch('odoo.addons.mail.models.mail_mail.send', patched_mail_mail_send):
    self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo().method_direct_trigger()
```

To try a bit, but I think you see my point.